### PR TITLE
Feature/uar 167:  retrieve beneficial owner private data call to "beneficial-owners" endpoint

### PR DESCRIPTION
**Pull request by olaych**
Added beneficial owners private data retrieval service from endpoint "/private/transactions/{transaction_id}/overseas-entity/{overseas_entity_id}/beneficial-owners"

### CODE REVIEW

File: src/services/overseas-entities/types.ts [hunk 1]
```diff
@@ -219,6 +219,31 @@ export interface BeneficialOwnerGovernmentOrPublicAuthorityResource {
     non_legal_firm_members_nature_of_control_types?: NatureOfControlType[];
 }
 
+export interface BeneficialOwnerPrivateDataResource {
+    pscId?: string,
```
Comment (azafarCH) [inline src/services/overseas-entities/types.ts:223]: I see the confluence page has the fields like this, but should we keep consistency with the rest of the interfaces with lowercase and underscores? Might have to ask someone

File: src/services/overseas-entities/types.ts [hunk 2]
```diff
@@ -219,6 +219,31 @@ export interface BeneficialOwnerGovernmentOrPublicAuthorityResource {
     non_legal_firm_members_nature_of_control_types?: NatureOfControlType[];
 }
 
+export interface BeneficialOwnerPrivateDataResource {
+    psc_id?: string,
+    date_became_registrable?: string,
+    date_of_birth?: InputDate,
```
Comment (azafarCH) [inline src/services/overseas-entities/types.ts:225]: I think this might have to be string, the other resources have string

File: src/services/overseas-entities/types.ts [hunk 3]
```diff
@@ -219,6 +219,31 @@ export interface BeneficialOwnerGovernmentOrPublicAuthorityResource {
...
+}
+
+export interface BeneficialOwnersPrivateDataResource {
+    bo_private_data?: BeneficialOwnerPrivateDataResource[];
+}
+
+export interface privateBoADDRESS {
+        address_line_1?: string,
+        address_line_2?: string,
+        care_of?: string,
```
Comment (bbell-companieshouse) [inline src/services/overseas-entities/types.ts:238]: Both care of and po box can be taken out as they are not used

File: src/services/overseas-entities/types.ts [hunk 4]
```diff
@@ -219,6 +219,31 @@ export interface BeneficialOwnerGovernmentOrPublicAuthorityResource {
...
+    is_service_address_same_as_usual_address?: string,
+    usual_residential_address?: privateBoADDRESS,
+    principal_address?: privateBoADDRESS,
+}
+
+export interface BeneficialOwnersPrivateDataResource {
+    bo_private_data?: BeneficialOwnerPrivateDataResource[];
+}
+
+export interface privateBoADDRESS {
```
Comment (bbell-companieshouse) [inline src/services/overseas-entities/types.ts:235]: Should we even call this PrivateAddress? Since we will use this format across BO + MO

File: src/services/overseas-entities/types.ts [hunk 5]
```diff
@@ -219,6 +219,56 @@ export interface BeneficialOwnerGovernmentOrPublicAuthorityResource {
...
+    bo_private_data: BeneficialOwnerPrivateDataResource[];
+}
+
+export interface BeneficialOwnersPrivateData {
+    boPrivateData: BeneficialOwnerPrivateData[];
+}
+
+export interface BeneficialOwnerPrivateData {
+    id?: string,
+    hashedId: string,
```
Comment (AndreKramer7654) [inline src/services/overseas-entities/types.ts:242]: as everything else is optional - can this be made optional too?

File: src/services/overseas-entities/types.ts [hunk 6]
```diff
@@ -219,6 +219,56 @@ export interface BeneficialOwnerGovernmentOrPublicAuthorityResource {
...
+    dateBecameRegistrable?: string,
+    dateOfBirth?: string,
+    isServiceAddressSameAsUsualAddress?: string,
+    usualResidentialAddress?: privateADDRESS,
+    principalAddress?: privateADDRESS,
+}
+
+export interface privateADDRESS {
```
Comment (AndreKramer7654) [inline src/services/overseas-entities/types.ts:250]: PrivateAddress

File: src/services/overseas-entities/types.ts [hunk 7]
```diff
@@ -219,6 +219,56 @@ export interface BeneficialOwnerGovernmentOrPublicAuthorityResource {
...
+        addressLine1?: string,
+        addressLine2?: string,
+        country?: string,
+        locality?: string,
+        postCode?: string,
+        premises?: string,
+        region?: string
+}
+
+export interface privateAddressResource {
```
Comment (AndreKramer7654) [inline src/services/overseas-entities/types.ts:260]: PrivateAddressRersource (pascal case)

File: test/services/company-psc/service.spec.ts [hunk 8]
```diff
@@ -1,7 +1,5 @@
 import chai from "chai";
 import sinon from "sinon";
-import chaiAsPromised from "chai-as-promised";
```
Comment (AndreKramer7654) [inline test/services/company-psc/service.spec.ts:3]: Did you mean to remove this?

File: src/services/overseas-entities/types.ts [hunk 9]
```diff
@@ -219,6 +219,56 @@ export interface BeneficialOwnerGovernmentOrPublicAuthorityResource {
     non_legal_firm_members_nature_of_control_types?: NatureOfControlType[];
 }
 
+export interface BeneficialOwnerPrivateDataResource {
+    id?: string,
+    hashed_id?: string,
+    date_became_registrable?: string,
```
Comment (AndreKramer7654) [inline src/services/overseas-entities/types.ts:225]: I see the other PR for MOs does not have everything as optional. Should we do the same and only have the id as optional?

File: src/services/overseas-entities/types.ts [hunk 10]
```diff
@@ -219,6 +219,56 @@ export interface BeneficialOwnerGovernmentOrPublicAuthorityResource {
...
+    id?: string,
+    hashedId?: string,
+    dateBecameRegistrable?: string,
+    dateOfBirth?: string,
+    isServiceAddressSameAsUsualAddress?: string,
+    usualResidentialAddress?: PrivateADDRESS,
+    principalAddress?: PrivateADDRESS,
+}
+
+export interface PrivateADDRESS {
```
Comment (bbell-companieshouse) [inline src/services/overseas-entities/types.ts:250]: Could we call this 'PrivateAddress' instead?

File: test/services/overseas-entities/overseas.entities.mock.ts [hunk 11]
```diff
@@ -47,6 +49,29 @@ export const ADDRESS: Address = {
     postcode: "BY 2"
 };
 
+export const privateADDRESS = {
```
Comment (bbell-companieshouse) [inline test/services/overseas-entities/overseas.entities.mock.ts:52]: PRIVATE_ADDRESS instead?

File: src/services/overseas-entities/service.ts [hunk 12]
```diff
@@ -88,4 +91,28 @@ export default class OverseasEntityService {
 
         return { httpStatusCode: resp.status };
     }
+
+    /**
+     * Get private beneficial owner data for an overseasentity
```
Comment (n-r-b) [inline src/services/overseas-entities/service.ts:96]:
```suggestion
     * Get private beneficial owner data for an overseas entity
```

File: src/services/overseas-entities/service.ts [hunk 13]
```diff
@@ -88,4 +91,28 @@ export default class OverseasEntityService {
...
+    public async getBeneficialOwnerPrivateData (transactionId: string, overseasEntityId: string): Promise< Resource<BeneficialOwnersPrivateData> | ApiErrorResponse > {
+        const URL = `private/transactions/${transactionId}/overseas-entity/${overseasEntityId}/beneficial-owners`
+        const response: HttpResponse = await this.client.httpGet(URL);
+
+        if (response.error) {
+            return {
+                httpStatusCode: response.status,
+                errors: [response.error]
+            };
+        };
```
Comment (n-r-b) [inline src/services/overseas-entities/service.ts:109]:
```suggestion
        }
```

File: src/services/overseas-entities/service.ts [hunk 14]
```diff
@@ -88,4 +91,28 @@ export default class OverseasEntityService {
...
+
+        const resource: Resource<BeneficialOwnersPrivateData> = {
+            httpStatusCode: response.status,
+            resource: Mapping.camelCaseKeys<BeneficialOwnersPrivateData>(response.body as BeneficialOwnersPrivateDataResource)
```
Comment (n-r-b) [inline src/services/overseas-entities/service.ts:113]: What's the reason for the camel case mapping? I noticed for the other private data retrieval endpoint, we keep this simple and do the following:

`email_address: (body.email_address)`

File: test/services/overseas-entities/overseas.entities.spec.ts [hunk 15]
```diff
@@ -495,4 +522,33 @@ describe("Mapping OverseasEntity Tests suite", () => {
 
         expect(dataResource.email_address).to.equal(undefined);
     });
+
+    it("maps private beneficial owner data fields correctly", async () => {
```
Comment (n-r-b) [inline test/services/overseas-entities/overseas.entities.spec.ts:526]: I don't think this test case belongs here under "Mapping OverseasEntity Tests suite" as the tests in this section cover functions in the `mapping.ts` class which we don't have. I feel we've already covered the mapping in the "should return httpStatusCode 200 for getBeneficialOwners method" test case, what do you think?

File: test/services/overseas-entities/overseas.entities.spec.ts [hunk 16]
```diff
@@ -147,6 +149,31 @@ describe("OverseasEntityService GET Tests suite", () => {
...
+    it("should return httpStatusCode 200 for getBeneficialOwners method", async () => {
+        sinon.stub(mockValues.requestClient, "httpGet").resolves(mockValues.mockBeneficialOwnerPrivateDataResponse[200]);
+
+        const oeService = new OverseasEntityService(mockValues.requestClient);
+        const data = (await oeService.getBeneficialOwnerPrivateData(
+            mockValues.TRANSACTION_ID,
+            mockValues.OVERSEAS_ENTITY_ID
+        )) as Resource<BeneficialOwnersPrivateData>;
+        expect(data.httpStatusCode).to.equal(200);
+        expect(data.resource).to.deep.equal(mockValues.BENEFICIAL_OWNER_PRIVATE_DATA_RESOURCE_MOCK);
```
Comment (n-r-b) [inline test/services/overseas-entities/overseas.entities.spec.ts:162]: We could add a test for the mapping of optional/empty values too?

File: src/services/overseas-entities/service.ts (hunk 14 above)
Comment (AndreKramer7654) [reply src/services/overseas-entities/service.ts:113]: The API generally tends to be using snake case (some_thing_like_this) and the Javascript expects camel case. Usually we convert by hand and not use the camelCaseKeys conversion as there are just a few fields and good to be explicit. We may have missed changing the email address but no big deal and I'm fine with the just convert everything some other services use. It lets through all new fields if we every add any without requiring changes.

File: test/services/overseas-entities/overseas.entities.spec.ts (hunk 16 above)
Comment (AndreKramer7654) [reply test/services/overseas-entities/overseas.entities.spec.ts:162]: If we just convert snake to camel case then not much point in further tests I'd say.

File: src/services/overseas-entities/service.ts (hunk 14 above)
Comment (n-r-b) [reply src/services/overseas-entities/service.ts:113]: Ok, that makes sense

File: src/services/overseas-entities/types.ts (hunk 1 above)
Comment (n-r-b) [reply src/services/overseas-entities/types.ts]: See the conversation here https://github.com/companieshouse/api-sdk-node/pull/597#discussion_r1314817261

**Review by n-r-b — APPROVED** [review]

**Review by AndreKramer7654 — APPROVED** [review]
